src/witness/circom.rs

```diff
@@ -49,38 +54,43 @@ impl Circom for Wasm {
 }
 
 #[cfg(feature = "circom-2")]
-impl Circom2 for Wasm {
+impl Circom2 for WasmInstance {
     fn get_field_num_len32(&self) -> Result<u32> {
         self.get_u32("getFieldNumLen32")
     }
 
     fn get_raw_prime(&self) -> Result<()> {
         let func = self.func("getRawPrime");
-        func.call(&[])?;
+        let mut store = self.store.write().unwrap();
+        func.call(&mut store, &[])?;
         Ok(())
     }
 
     fn read_shared_rw_memory(&self, i: u32) -> Result<u32> {
         let func = self.func("readSharedRWMemory");
-        let result = func.call(&[i.into()])?;
+        let mut store = self.store.write().unwrap();
```

Comment:
Why not?
```suggestion
        let mut store = self.store.read().unwrap();
```
Same for all other "getters"

Reply:
Good eye. Sorry for the overtly long response, but I just dove into this because I was curious as well.

Decompiling `circom2_multiplier2.wasm`, we will see the following statements:

```wat
  ;; memory initialization and exports
  (memory (;0;) 1008)
  (export "memory" (memory 0))
  ;; function exports
  (export "getVersion" (func 84))
   ...
  (export "readSharedRWMemory" (func 86)) ;; this is what we want!
  (export "writeSharedRWMemory" (func 87)) ;; this is what we want!
  ...
  ;; more function exports;
```

We can see that readSharedRWMemory doesn't touch memory (86), but writeSharedRWMemory does (87):

```wat
  (func (;86;) (type 14) (param i32) (result i32)
    i32.const 1992 ;; this is where the shared RW memory address must start!
    local.get 0
    i32.const 4
    i32.mul
    i32.add
    i32.load) ;; READ
  (func (;87;) (type 16) (param i32 i32)
    i32.const 1992
    local.get 0
    i32.const 4
    i32.mul
    i32.add
    local.get 1
    i32.store) ;; WRITE
```

So that begs the question, why can't we just readlock the store if we're only reading?

Turns out, the runtime has no clue if a function modifies memory, and therefore, it must writelock the [Store](https://webassembly.github.io/spec/core/exec/runtime.html#store) (which stores the actual memory) anytime we want to call a Wasm function.

This makes an interesting point:, if we were reading that same `Memory` from `SafeMemory` instead of the runtime (remember, we are sharing `Memory` via `Arc<RwLock<Store>>` we could get away with just readlocking it.

Now here's the kicker: because we just reverse-engineering `readSharedRWMemory` (and `writeSharedRWMemory`), we will see there is almost an equivalent function in `SafeMemory` already: `read_u32` and `write_u32`. We could replace the `WasmInstance`'s `fn read_shared_rw_memory` with `SafeMemory`'s `read_u32`, assuming we handle the u32 as i32 (2's complement) and handle the base address offset (by adding 1992 to the virtual pointer, which is what the  `i32.const 1992` hints at in the above `Wasm` functions)